drivers/gpio/gpio_cy8c95xx.c

```diff
@@ -0,0 +1,329 @@
+
+/** Cache of the output configuration and data of the pins. */
+struct cy8c95xx_pin_state {
+	uint16_t dir;
+	uint16_t data;
+} __packed;
+
+/** Runtime driver data */
+struct cy8c95xx_drv_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	const struct device *i2c_master;
+	struct cy8c95xx_pin_state pin_state;
+	struct k_sem lock;
+};
+
+/** Configuration data */
+struct cy8c95xx_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	const char *i2c_master_dev_name;
+	uint16_t i2c_slave_addr;
+};
+
+#define CY8C95XX_REG_INPUT_DATA0                0x00
+#define CY8C95XX_REG_INPUT_DATA1                0x01
+#define CY8C95XX_REG_INPUT_DATA2                0x02
+#define CY8C95XX_REG_OUTPUT_DATA0               0x08
+#define CY8C95XX_REG_OUTPUT_DATA1               0x08
+#define CY8C95XX_REG_OUTPUT_DATA2               0x08
+
+#define CY8C95XX_REG_PORT_SELECT                0x18
+#define CY8C95XX_REG_DIR                        0x1C
+#define CY8C95XX_REG_PULL_UP                    0x1D
+#define CY8C95XX_REG_PULL_DOWN                  0x1E
+#define CY8C95XX_REG_ID                         0x2E
+
+static int write_pin_state(const struct cy8c95xx_config *cfg,
+			   struct cy8c95xx_drv_data *drv_data,
+			   struct cy8c95xx_pin_state *pins)
+{
+	uint8_t dir[3] = { (pins->dir & 0xFF), ((pins->dir >> 4) & 0xF0),
+			   ((pins->dir >> 12) & 0x0F) };
+	uint8_t data[3] = { (pins->data & 0xFF), ((pins->data >> 4) & 0xF0),
+			    ((pins->data >> 12) & 0x0F) };
+	int rc;
+
+	for (int8_t i = 0; i < 3; i++) {
+		rc = i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+					CY8C95XX_REG_PORT_SELECT, i);
+		if (rc == 0) {
+			rc = i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+						CY8C95XX_REG_DIR, dir[i]);
+		}
+	}
+	rc = i2c_burst_write(drv_data->i2c_master, cfg->i2c_slave_addr,
+			     CY8C95XX_REG_OUTPUT_DATA0, data, 3);
+
+	return rc;
+}
+
+static int cy8c95xx_config(const struct device *dev,
+			   gpio_pin_t pin,
+			   gpio_flags_t flags)
+{
+	const struct cy8c95xx_config *cfg = dev->config;
+	struct cy8c95xx_drv_data *drv_data = dev->data;
+	struct cy8c95xx_pin_state *pins = &drv_data->pin_state;
+	int rc = 0;
+	uint32_t io_flags;
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	/* No strengths support */
+	if ((flags & (GPIO_DS_ALT_LOW | GPIO_DS_ALT_HIGH)) != 0) {
+		return -ENOTSUP;
+	}
+
+	/* No open-drain support */
+	if ((flags & GPIO_SINGLE_ENDED) != 0U) {
+		return -ENOTSUP;
+	}
+
+	/* Does not support pull-up/pull-down */
+	if ((flags & (GPIO_PULL_UP | GPIO_PULL_DOWN)) != 0U) {
+		return -ENOTSUP;
+	}
+
+	/* Does not support disconnected pin, and
+	 * not supporting both input/output at same time.
+	 */
+	io_flags = flags & (GPIO_INPUT | GPIO_OUTPUT);
+	if ((io_flags == GPIO_DISCONNECTED)
+	    || (io_flags == (GPIO_INPUT | GPIO_OUTPUT))) {
+		return -ENOTSUP;
+	}
```

Comment:
According to the documentation the chip supports simultaneous read/write, Hi-Z pin state. It would be nice to have at least in/out mode supported as it makes it easier to test the driver.

Reply:
I have some misunderstanding, it can be configured pull up/down per pin, I have added and tested. it works.

it also supports simultaneous read/write, when configured as output, I can read the pin and get the right value.

Reply:
I believe the device supports some more configurable features as well, like drive strength and open drain. Please either add the support (doesn't seem too complex), or mention in the comments that the driver doesn't (yet) support these features of the device.